src/utilities.jl

```diff
@@ -145,3 +145,69 @@ end
 
 check_positive_definite(M::AbstractMatrix; kwargs...) =
   check_positive_definite(LinearOperator(M); kwargs...)
+
+
+  """
+  solve_shifted_system!(op::LBFGSOperator{T,I,F1,F2,F3}, 
+                        z::AbstractVector{T}, 
+                        σ::T, 
+                        γ_inv::T, 
+                        inv_Cz::AbstractVector{T}, 
+                        p::AbstractVector{T}, 
+                        v::AbstractVector{T}, 
+                        u::AbstractVector{T}) where {T,I,F1,F2,F3}
+
+Solves linear system (B + σI) x = b, where B is a forward L-BFGS operator and σ <= 0.
+
+### Parameters
+
+- `op::LBFGSOperator{T,I,F1,F2,F3}`: The L-BFGS operator `B_k`. Encodes the curvature information used to approximate the Hessian.
+- `z::AbstractVector{T}`: The vector representing `-b`.
+- `σ::T`: Nonnegative shift.
+- `inv_Cz::AbstractVector{T}`: A preallocated vector used to store the result of the solution. It will be overwritten in the function to hold the computed `s`.
+
+### Returns
+
+- `inv_Cz::AbstractVector{T}`: The solution vector `s` such that `(B_k + σ * I) s = -∇f(x_k)`.
+
+### Method
+
+The method uses a two-loop recursion-like approach with modifications to handle the shift `σ`.
+
+### References
+Erway, J. B., Jain, V., & Marcia, R. F. (2013). Shifted L-BFGS Systems.
+"""
+
+function solve_shifted_system!(
+  op::LBFGSOperator{T, I, F1, F2, F3},
+  z::AbstractVector{T},
+  σ::T,
+  inv_Cz::AbstractVector{T},
+  ) where {T, I, F1, F2, F3}
+  data = op.data
+  insert = data.insert
+  
+  γ_inv = 1 / data.scaling_factor
+  inv_c0 = 1 / (γ_inv + σ)
+  @. inv_Cz = inv_c0 * z
+
+  max_i = 2 * data.mem
+  for i = 1:max_i
+    j = (i + 1) ÷ 2
+    k = mod(insert + j - 1, data.mem) + 1
+    data.shifted_u .= ((i % 2) == 0 ? data.b[k] : data.a[k])
+
+    @. data.shifted_p[:, i] = inv_c0 * data.shifted_u
+
+    for t = 1:(i - 1)
+      c0 = dot(view(data.shifted_p, :, t), data.shifted_u)
+      c1= ((t % 2) == 0 ? -1 : 1) .*data.shifted_v[t]
```

Comment:
We discussed not computing `t %2` every time.

Reply:
I thought you wanted me to use this instead of (-1)^t?

Reply:
@dpo  what do you suggest?

Reply:
I suggested using a flag 
check on the new code